File: examples/C#/rtpapa.cs
//
//  Custom routing Router to Papa (ROUTER to REP)
//

//  Author:     Michael Compton, Tomas Roos
//  Email:      michael.compton@littleedge.co.uk, ptomasroos@gmail.com

using System.Text;
using System.Threading;
using ZMQ;

namespace ZMQGuide
{
    internal class Program
    {
        //  We will do this all in one thread to emphasize the sequence
        //  of events...
        public static void Main(string[] args)
        {
            using (var context = new Context(1))
            {
                using (Socket client = context.Socket(SocketType.ROUTER), worker = context.Socket(SocketType.REP))
                {
                    client.Bind("inproc://routing");

                    worker.StringToIdentity("A", Encoding.Unicode);
                    worker.Connect("inproc://routing");

                    //  Wait for the worker to connect so that when we send a message
                    //  with routing envelope, it will actually match the worker...
                    Thread.Sleep(1000);

                    //  Send papa address, address stack, empty part, and request
                    client.SendMore("A", Encoding.Unicode);
                    client.SendMore("address 3", Encoding.Unicode);
                    client.SendMore("address 2", Encoding.Unicode);
                    client.SendMore("address 1", Encoding.Unicode);
                    client.SendMore("", Encoding.Unicode);
                    client.Send("This is the workload", Encoding.Unicode);

                    //  Worker should get just the workload
                    ZHelpers.Dump(worker, Encoding.Unicode);

                    //  We don't play with envelopes in the worker
                    worker.Send("This is the reply", Encoding.Unicode);

                    //  Now dump what we got off the ROUTER socket...
                    ZHelpers.Dump(client, Encoding.Unicode);
                }
            }
        }
    }
}

File: examples/Scala/rtpapa.scala
/*
 *  Custom routing Router to Papa (ROUTER to REP)
 *
 *
 *  @Author:     Giovanni Ruggiero
 *  @Email:      giovanni.ruggiero@gmail.com
 */

import org.zeromq.ZMQ
import ZHelpers._

object rtpapa  {
  def main(args : Array[String]) {
		val NOFLAGS = 0
		//  We will do this all in one thread to emphasize the sequence
		//  of events...
		val ctx = ZMQ.context(1)
		val client = ctx.socket(ZMQ.ROUTER)
		val worker = ctx.socket(ZMQ.REP)

		client.bind("inproc://routing")
		worker.setIdentity("A".getBytes)
		worker.connect("inproc://routing")

		//  Wait for the worker to connect so that when we send a message
		//  with routing envelope, it will actually match the worker...
		Thread.sleep(1000)

		//  Send papa address, address stack, empty part, and request
		client.send("A".getBytes, ZMQ.SNDMORE)
		client.send("address 3".getBytes, ZMQ.SNDMORE)
		client.send("address 2".getBytes, ZMQ.SNDMORE)
		client.send("address 1".getBytes, ZMQ.SNDMORE)
		client.send("".getBytes, ZMQ.SNDMORE)
		client.send("This is the workload".getBytes, NOFLAGS)

		//  Worker should get just the workload
		dump(worker)

		//  We don't play with envelopes in the worker
		worker.send("This is the reply".getBytes, NOFLAGS)

		//  Now dump what we got off the ROUTER socket...
		dump(client)
	}
}
